File: src/main/java/me/testaccount666/serversystem/userdata/persistence/KitMapFieldHandler.java
package me.testaccount666.serversystem.userdata.persistence;

import me.testaccount666.serversystem.ServerSystem;
import me.testaccount666.serversystem.commands.executables.kit.manager.KitManager;
import org.bukkit.configuration.file.FileConfiguration;

import java.util.HashMap;
import java.util.Map;

public class KitMapFieldHandler implements FieldHandler<Map<String, Long>> {
    @Override
    public void save(FileConfiguration config, String path, Map<String, Long> value) {
        config.set(path, null);
        if (value == null || value.isEmpty()) return;
        var kitManager = ServerSystem.Instance.getRegistry().getService(KitManager.class);
        if (kitManager == null) return;

        value.forEach((kitName, cooldown) -> {
            if (!kitManager.kitExists(kitName)) return;
            config.set("${path}.${kitName}", cooldown);
        });
    }

    @Override
    public Map<String, Long> load(FileConfiguration config, String path, Map<String, Long> defaultValue) {
        if (!config.isSet(path)) return defaultValue;
        if (ServerSystem.Instance.getRegistry().getService(KitManager.class) == null) return defaultValue;

        var kitCooldowns = new HashMap<String, Long>();
        config.getConfigurationSection(path).getKeys(false).forEach(kitName -> kitCooldowns.put(kitName, config.getLong("${path}.${kitName}")));
        return kitCooldowns;
    }
}

File: src/main/kotlin/me/testaccount666/serversystem/userdata/persistence/KitMapFieldHandler.kt
package me.testaccount666.serversystem.userdata.persistence

import me.testaccount666.serversystem.ServerSystem
import me.testaccount666.serversystem.commands.executables.kit.manager.KitManager
import me.testaccount666.serversystem.extensions.getServiceOrNull
import org.bukkit.configuration.file.FileConfiguration

class KitMapFieldHandler : FieldHandler<HashMap<String, Long>> {
    override fun save(config: FileConfiguration, path: String, value: HashMap<String, Long>?) {
        config[path] = null
        if (value.isNullOrEmpty()) return
        val kitManager = getServiceOrNull<KitManager>() ?: return

        value.forEach { (kitName: String, cooldown: Long) ->
            if (!kitManager.kitExists(kitName)) return@forEach
            config["${path}.${kitName}"] = cooldown
        }
    }

    override fun load(config: FileConfiguration, path: String, defaultValue: HashMap<String, Long>?): HashMap<String, Long>? {
        if (!config.isSet(path)) return defaultValue
        if (!ServerSystem.instance.registry.hasService<KitManager>()) return defaultValue

        val kitCooldowns = HashMap<String, Long>()
        config.getConfigurationSection(path)!!.getKeys(false).forEach { kitCooldowns[it] = config.getLong("${path}.${it}") }
        return kitCooldowns
    }
}